### ISSUE TITLE
LazyComboBox implementations

### ISSUE THREAD

**Issue by lorenzobenvenuti** [viritin/viritin#66]
Hi Matti,

I played a bit with the LazyComboBox implementation and IMHO there are a couple of features that could make it less resource intensive:

- Introduce something similar to the TextField.setTextChangeEventMode to choose how often the combo should filter the data; I noticed that right now the count (and possibly filterEntities) are called more or less for every keystroke. I'm not sure how easy it could be, the VFilterSelect/ComboBox API don't seem so extension-friendly.

- With real large datasets the method setValue triggers a lot of calls to the filterEntities to "search" the current element (expecially if the value is one of the last in the dataset). It would be nice to set an (optional?) IndexProvider to allow the component to load only the page where the current value is stored. I'm thinking to something like

```java
interface IndexProvider<T> {

    int getEntityIndex(T entity, String filter);

}
```

Thank you and keep up the good work!

lorenzo

**Comment by mstahv** [comment]
Hi,

Good points. The "eagerness" of ComboBox should definitely be configurable. ATM I think it has some very small value. Would you create an issue about it to dev.vaadin.com and cross reference it both ways?

The ComboBox does have a method to bypass random indexOf calls to container (setScrollToSelectedItem(boolean)), but when I was developing the LazyComboBox I think I couldn't get it work in any sane way :-( I'm afraid there is some regression in the core for this feature. Because it didn't work I created "item index cache" that works in simple cases, but that is only populated when actually scrolling to some page. That works decently when choosing new values, but it can never work well with all cases (e.g. preset value on some n+1 page).

If you have time too look at if setScrollToSelectedItem(false) could be made to work it would be awesome. If it is broken, again, a ticket to dev.vaadin.com is the way to get things forward. I can also try to look at this, but I'll be on a skiing trip for the next week so don't hold your breath.

**Comment by lorenzobenvenuti** [comment]
Hi,

I opened the ticket about the ComboBox eagerness: http://dev.vaadin.com/ticket/17137.

Regarding the setScrollToSelectedItem(false) I tried it, but the findEntities method is still invoked a lot of times because the AbstractSelect.setValue checks if the value is valid using the method Container.containsId. I guess you can't avoid this, it's the only thing the Combo can do to check the input; the "IndexProvider" I suggested would allow a more efficient implementation of the containsId for large datasets.

Enjoy you skiing trip!

Thank you,

lorenzo

**Comment by mstahv** [comment]
setScrollToSelectedItem is now fixed in 7.5.0, but looks like it wasn't the only issue that limits the scalability of ComboBox. It appears that AbstractSelect also verifies the existence of item using containsId when setting the value -> still need to scroll the whole resultset.

**Comment by mstahv** [comment]
If you have a good real world test case for combobox with large dataset, would you test with the current master branch and Vaadin 7.5 if it now works decently?

**Comment by lorenzobenvenuti** [comment]
I hope I haven't deleted the project I used to play with LazyComboBox; anyway I should be able  to rewrite the code in a few minutes. I'll let you know.

**Comment by lorenzobenvenuti** [comment]
Hi,

I finally had some time to test my code with the current version of Vaadin and Viritin. To summarize:

- Previous results (Vaadin 7.4.1, Viritin 1.29, scrollToSelectedItem set to false): when lazyComboBox.setValue is invoked the items are "scrolled" until the current value is found and nothing is shown on the combo
- Current result (Vaadin 7.5.1, Viritin master branch, scrollToSelectedItem set to false): the items are not scrolled (yay) but still no selection is shown on the combo. Is there anyway to update the displayed value?

Another thing: with Viritin master branch (and in version 1.35 too) if I add a ValueChangeListener to LazyComboBox it gets called twice; it seems to work ok in my previous project (Viritin 1.29).

Thank you,

lorenzo

**Comment by cbmeeks** [comment]
Has there been any progress on this?

I am having the same issue.  I am using Vaadin 7.5.5 and Viritin 1.35.

We have a LazyComboBox that is backed by about 185 records.  Not much.  But, when user types in something from page 2 onward and tabs away, clicking the combo again causes the value to disappear and it cannot find the original value.  So the user has to find it again.

**Comment by mstahv** [comment]
Hi, 7.5.5. + Viritin 1.35 should behave "properly". The ComboBox just doesn't scroll to the "page" that contains the selected item, as it might be really expensive operation. If you click somewhere else the old selection should be intact.

**Comment by sebdavid** [comment]
Hi,
I'm having the same issue than cbmeeks, with Vaadin 7.5.7 + Viritin 1.37.
There are the steps :
1. the LazyComboBox is displayed with no selected value
2. open the combobox, scroll to another page than the first, select an item
3. the selected item is displayed 
4. open the combobox : the displayed text is cleared
5. click somewhere else : the combobox is closed, and the text of the original selection is not displayed

But, it seems that the combobox still has the value. It is just the displayed text that is not there.
I think I'm not doing anything particular. I have a custom class extending the LazyComboBox, and the following constructor :
```java
public abstract class AbstractLazyCB<T> extends LazyComboBox<T>  {

	protected AbstractLazyCB(){
		super();

		// here are some initialization
		...
		
		// init the CB
		initList(
				type,
				anotherClass::search, 
				anotherClass::count);
	}
}
```

**Comment by mstahv** [comment]
I investigated this and to me it seems like an other bug in the "scalable mode" in cores ComboBox. I created an issue for it here:
https://dev.vaadin.com/ticket/19221#ticket

I have a patch almost done as well, but the combination of jet lag and Vaadin cores selects code is a bad combination - I'll need to review my stuff tomorrow.

**Comment by mstahv** [comment]
Finally the fix in the core is published with 7.7!
